### PR TITLE
Rewrite tracert

### PR DESCRIPTION
**Pull request by gedmurphy**
Our tracert was written before we had ICMP support, therefore it manually crafted ping requests using raw sockets. It also didn't run on newer versions of windows, and had hard-coded text.
This new version now uses the new ICMP APIs, runs on win10 and has multi-language support. It also brings the code somewhat into the 21st century.

I've still got a couple of bits to cleanup, so I'll merge this PR next time I get a spare couple of hours.

### CODE REVIEW

File: base/applications/network/tracert/tracert.rc [hunk 1]
```diff
@@ -1,5 +1,65 @@
...
+#ifdef LANGUAGE_UK_UA
+#include "lang/uk-UA.rc"
+#endif
+#ifdef LANGUAGE_ZH_CN
+#include "lang/zh-CN.rc"
+#endif
+#ifdef LANGUAGE_ZH_TW
+#include "lang/zh-TW.rc"
+#endif
+#endif
```
Comment (HBelusca) [inline base/applications/network/tracert/tracert.rc]: Please add a trailing empty newline.

File: base/applications/network/tracert/tracert.cpp [hunk 2]
```diff
@@ -398,32 +399,32 @@ RunTraceRoute()
...
-                                        NULL,
-                                        &Source,
-                                        (struct sockaddr_in6 *)Info.Target->ai_addr,
-                                        SendBuffer,
-                                        (USHORT)PACKET_SIZE,
-                                        &IpOptionInfo,
-                                        ReplyBuffer,
-                                        ReplySize,
-                                        Info.Timeout);
+                (void)Icmp6SendEcho2(Info.hIcmpFile,
```
Comment (HBelusca) [inline base/applications/network/tracert/tracert.cpp]: There is no need to check the return value of these calls?

File: base/applications/network/tracert/CMakeLists.txt [hunk 3]
```diff
@@ -1,6 +1,5 @@
 
 add_executable(tracert tracert.cpp tracert.rc)
 set_module_type(tracert win32cui UNICODE)
-target_link_libraries(tracert conutils ${PSEH_LIB})
```
Comment (HBelusca) [inline base/applications/network/tracert/CMakeLists.txt]: Removed on purpose?

File: base/applications/network/tracert/tracert.cpp [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+        {
+            StringCchCopyW(Info.HostName, NI_MAXHOST, argv[i]);
+            break;
+        }
+    }
+
+    return true;
+}
+
+int wmain(int argc, wchar_t *argv[])
```
Comment (learn-more) [inline base/applications/network/tracert/tracert.cpp]: Add `extern "C"` or `EXTERN_C` to fix the gcc build.

Comment (gedmurphy) [reply base/applications/network/tracert/tracert.cpp]: Thanks 👍

**Comment by HBelusca** [comment]
You also need to definitively fix your name in the patch author :p

**Comment by HBelusca** [comment]
Ok I've fixed the author name, but please use a full name instead of a nickname for the next times!
I've also squashed together the encoding-fix commit and the fix-GCC-build commits together with the first one.

File: base/applications/network/tracert/tracert.rc [hunk 5]
```diff
@@ -1,5 +1,61 @@
...
-#define REACTOS_STR_ORIGINAL_COPYRIGHT "Ged Murphy (gedmurphy@gmail.com)"
+#define REACTOS_STR_ORIGINAL_COPYRIGHT "Ged Murphy (gedmurphy@reactos.org)"
 #include <reactos/version.rc>
+
+
+/* UTF-8 */
+#pragma code_page(65001)
+
+#ifdef LANGUAGE_BG_BG
+#include "lang/bg-BG.rc"
```
Comment (ThFabba) [inline base/applications/network/tracert/tracert.rc]: Can you indent these by 4 spaces? All the other modules do it that way.

File: base/applications/network/tracert/tracert.cpp [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+static ULONG
+GetULONG(
+    _In_z_ LPWSTR String
+)
+{
+    ULONG Length;
+    Length = wcslen(String);
+
+    ULONG i = 0;
+    while ((i < Length) && ((String[i] < L'0') || (String[i] > L'9'))) i++;
```
Comment (ThFabba) [inline base/applications/network/tracert/tracert.cpp]: iswdigit would simplify these

**Pull request merged**